### PR TITLE
tooling: Async/pathlib/mypy cleanups and utils

### LINKED ISSUE #17487: Use python `pathlib` more and clean up some existing code with it
### description

i have a bit of an aversion to using pathlib (which i need to get over)  - i guess because im used to `os` 

ive landed quite a bit of code where `pathlib` would probs be preferrable

this is a ticket to track cleaning it up

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: tooling: Improve async runner/test handling
Additional Description:

This Pr does a few things:

- adds an `async_property` decorator which is v handy for writing async classes
- switches pytest to `-Werror` which is kinda necessary for catching asyncio unawaited errors.
- adds a dep pin for setuptools to base/utils - some environments were still using older versions of setuptools which can cause issues with `-Werror` (as well as other incompatibilities)
- makes `runner.catches` work async
- improves the `main` plugin test to work async
- adds some tests for pytest plugin fixtures
- mypy fixes/cleanups
- `os` -> `pathlib` cleanups (Fix #17487 )

re the setuptools dep - i tried removing from docs requirements - but it wouldnt let me - setuptools works in mysterious ways

The decorator can be used as so:

```python

import asyncio

from envoy.tools.base.functional import async_property


Class MyClass(object):

    @async_property
    async def myprop(self):
        return await some_other_coroutine()

    async def run(self):
        print(await self.myprop)

asyncio.run(MyClass().run())
```
it can also cache the results of the async call like so:

```python

Class MyClass(object):

    @async_property(cache=True)
    async def myprop(self):
        return await some_other_coroutine()
```

in this case repeated awaits of the property will return the cached results.

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
coverage is here https://storage.googleapis.com/envoy-pr/17505/tooling/index.html

File: tools/base/functional.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            # Each time its called it needs to replace itself with a
+            # fresh coroutine version of itself, as the last one has
+            # been fired.
+            # The result stays the same though.
+            self._cache_result(instance, _result())
+            return result
+
+        if self.cache:
+            self._cache_result(instance, _result())
+        return result
```
Comment (phlax) [inline tools/base/functional.py:54]: this can just return `_result()`

File: tools/base/tests/test_functional.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+        assert (
+            e.value.args[0]
+            == 'AN ERROR OCCURRED')
+        assert (
+            list(m_async.call_args)
+            == [(), {}])
+        return
+
+    # results can repeatedly awaited
```
Comment (phlax) [inline tools/base/tests/test_functional.py:69]: can be

**Comment by phlax** [comment]
ive figured out a better implementation that doesnt depend on gc for closing coroutines

WIP while i reimplement...

File: tools/base/tests/test_functional.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            if raises:
+                await m_async()
+                raise SomeError("AN ERROR OCCURRED")
+            else:
+                return await m_async()
+
+    klass = Klass()
+
+    # The class.prop should be an instance of async_prop
+    # As `__get__` is async you have to `await` it
```
Comment (phlax) [inline tools/base/tests/test_functional.py:36]: this is wrong...

File: tools/base/functional.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    async def async_result(self) -> Any:
+        # retrieve the value from cache if available
+        cached_property = self.get_cached_prop()
+        if cached_property:
+            return cached_property
+
+        # derive the result, set the cache if required, and return the result
+        return self.set_prop_cache(await self.fun(self._instance))
+
+    def get_cached_prop(self) -> Optional[dict]:
```
Comment (phlax) [inline tools/base/functional.py:48]: `Optional[Any]`

File: tools/base/functional.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        return self
+
+    def __get__(self, instance: Any, cls=None) -> Any:
+        if instance is None:
+            return self
+        self._instance = instance
+        return self.async_result()
+
+    @property
+    def prop_cache(self):
```
Comment (phlax) [inline tools/base/functional.py:34]: `-> dict`

File: tools/base/tests/test_functional.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+import pytest
+
+from tools.base import functional
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("cache", [None, True, False])
+@pytest.mark.parametrize("raises", [True, False])
+async def test_functional_async_property(cache, raises):
```
Comment (phlax) [inline tools/base/tests/test_functional.py:11]: this is probs more of an integration test than a unit test but as its reaching into the bowels of python im kinda happy with that

File: tools/testing/plugin.py [hunk 7]
```diff
@@ -37,20 +38,40 @@ def _patches(*args, prefix: str = "") -> ContextManager:
 
 
 @pytest.fixture
-def patches():
+def patches() -> ContextManager:
```
Comment (phlax) [inline tools/testing/plugin.py:41]: `Type[ContextManager]`

File: tools/base/tests/test_runner.py [hunk 8]
```diff
@@ -54,9 +54,18 @@ def run(self, *args, **kwargs):
                 raise self.raises("AN ERROR OCCURRED")
             return result
 
+        @runner.catches(errors)
+        async def run_async(self, *args, **kwargs):
+            result = self._runner(*args, **kwargs)
+            if self.raises:
+                raise self.raises("AN ERROR OCCURRED")
```
Comment (htuch) [inline tools/base/tests/test_runner.py:61]: Is this the best we can do? No additional contextual information? Just wondering, as it doesn't seem super actionable on its own.

File: tools/dependency/tests/test_pip_check.py [hunk 9]
```diff
@@ -31,51 +31,68 @@ def test_pip_checker_config_requirements():
...
-    with patched as (m_utils, m_path, m_join):
-        assert checker.dependabot_config == m_utils.from_yaml.return_value
+    with patched as (m_utils, m_path):
+        if isdict:
+            m_utils.from_yaml.return_value = {}
+
+        if isdict:
+            assert checker.dependabot_config == m_utils.from_yaml.return_value
+        else:
+
```
Comment (htuch) [inline tools/dependency/tests/test_pip_check.py:49]: I'm thinking we're going to have to agree to disagree on this style, but each time I see it I find i very jarring..

File: tools/extensions/tests/test_extensions_check.py [hunk 10]
```diff
@@ -29,28 +30,54 @@ def test_extensions_checker_all_extensions():
...
+
+    with patched as (m_inst, m_spec, m_loader, m_module):
+        m_inst.side_effect = _is_instance
+
+        if is_module and is_loader:
+            assert (
+                checker.configured_extensions
+                == m_module.return_value.EXTENSIONS)
+        else:
+
```
Comment (htuch) [inline tools/extensions/tests/test_extensions_check.py:57]: :(

**Review by htuch — COMMENTED** [review]
Looks good, generally prefer much smaller PRs but in this case it seems fine as a cleanup rollup.

**Comment by phlax** [comment]
> generally prefer much smaller PRs but in this case it seems fine as a cleanup rollup.

yep, apologies, it kinda morphed

also, mypy can be a bit of rabbit hole of type safety, and changing only some of the paths to pathlib was making it unhappy (as do functions/methods that can return more than one type)

File: tools/dependency/tests/test_pip_check.py (hunk 9 above)
Comment (phlax) [reply tools/dependency/tests/test_pip_check.py:49]: k - i give in - ill remove the preceding ws from the contexts

it jars with me also, but at the same time ive read so much code that has a context buried deep in it that is not obvious what is going on

Comment (phlax) [reply tools/dependency/tests/test_pip_check.py:49]: (i guess also a dislike of ~old-js style nested closures)

File: tools/base/functional.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        # retrieve the value from cache if available
+        cached_property = self.get_cached_prop()
+        if cached_property:
+            return cached_property
+
+        # derive the result, set the cache if required, and return the result
+        return self.set_prop_cache(await self.fun(self._instance))
+
+    def get_cached_prop(self) -> Any:
+        if not self.cache:
```
Comment (phlax) [inline tools/base/functional.py:53]: this wont cache `None` atm which it probs should

File: tools/code_format/python_check.py [hunk 12]
```diff
@@ -44,30 +46,30 @@ def flake8_app(self) -> Flake8Application:
         return flake8_app
 
     @property
-    def flake8_args(self) -> list:
-        return ["--config", self.flake8_config_path, self.path]
+    def flake8_args(self) -> List[str]:
```
Comment (phlax) [inline tools/code_format/python_check.py:49]: this can be a tuple

File: tools/docs/sphinx_runner.py [hunk 13]
```diff
@@ -182,25 +188,32 @@ def check_env(self) -> None:
             raise SphinxEnvError(
                 "Given git tag does not match the VERSION file content:"
                 f"{self.docs_tag} vs v{self.version_number}")
-        with open(os.path.join(self.rst_dir, "version_history/current.rst")) as f:
-            if not self.version_number in f.read():
-                raise SphinxEnvError(
-                    f"Git tag ({self.version_number}) not found in version_history/current.rst")
+        # this should probs only check the first line
+        version_current = self.rst_dir.joinpath("version_history/current.rst").read_text()
```
Comment (phlax) [inline tools/docs/sphinx_runner.py:192]: joinpath parts

File: tools/base/tests/test_runner.py (hunk 8 above)
Comment (phlax) [reply tools/base/tests/test_runner.py:61]: comment added at the top of the dummy class

**Comment by phlax** [comment]
updated

as an aside @htuch writing/testing so many props is making reconsider not using `attrs` - that is what it kinda gives you free along with class/prop validation etc - and works out of the box with mypy

on my travels through the interwebs i read gvr saying somewhere that python was in the process of adopting some of the attrs features so i need to investigate this a bit further

**Review by htuch — APPROVED** [review]
Epic, thanks!

**Pull request merged**